extensions/wrapper/wrapper/src/main/java/de/sovity/edc/ext/wrapper/api/ui/pages/transferhistory/TransferHistoryPageDataFetcher.java

```diff
@@ -0,0 +1,136 @@
+import org.eclipse.edc.connector.spi.asset.AssetService;
+import org.eclipse.edc.connector.spi.contractagreement.ContractAgreementService;
+import org.eclipse.edc.connector.spi.transferprocess.TransferProcessService;
+import org.eclipse.edc.connector.transfer.spi.types.TransferProcess;
+import org.eclipse.edc.spi.entity.Entity;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.spi.types.domain.asset.Asset;
+import org.jetbrains.annotations.NotNull;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import static de.sovity.edc.ext.wrapper.utils.EdcDateUtils.utcMillisToOffsetDateTime;
+import static java.util.stream.Collectors.*;
+
+@RequiredArgsConstructor
+public class TransferHistoryPageDataFetcher {
+    private final AssetService assetService;
+    private final ContractAgreementService contractAgreementService;
+    private final ContractNegotiationStore contractNegotiationStore;
+    private final TransferProcessService transferProcessService;
+
+    private final TransferProcessStateService transferProcessStateService;
+
+    /**
+     * Fetches all contract agreements as {@link TransferHistoryEntry}s.
+     *
+     * @return {@link TransferHistoryEntry}s
+     */
+    @NotNull
+    public List<TransferHistoryEntry> getTransferHistoryEntries() {
+
+
+        var negotiationsByID = getAllContractNegotiations().stream()
+                .filter(it -> it.getContractAgreement() != null)
+                .collect(groupingBy(
+                        it -> it.getContractAgreement().getId(),
+                        collectingAndThen(maxBy(Comparator.comparing(Entity::getCreatedAt)), Optional::get)
+                ));
+
+
+        var agreementsById = getAllContractAgreements().stream().collect(toMap(
+                ContractAgreement::getId, Function.identity()
+        ));
+
+        var assetsById = getAllAssets().stream()
+                .collect(toMap(Asset::getId, Function.identity()));
+
+        var transferStream = getAllTransferProcesses().stream();
+
+        return transferStream.map(process -> {
+
+            var agreement =
+                    agreementsById.get(process.getDataRequest().getContractId());
+            var negotiation = negotiationsByID.get(process.getDataRequest().getContractId());
+            var asset = assetLookup(assetsById, process);
+            ContractAgreementDirection direction = ContractAgreementDirection.fromType(negotiation.getType());
+
+            var transferHistoryEntry = new TransferHistoryEntry();
+
+            transferHistoryEntry.setAssetId(asset.getId());
+
+            if (direction == ContractAgreementDirection.CONSUMING) {
+                transferHistoryEntry.setAssetName(asset.getId());
+            } else {
+                transferHistoryEntry.setAssetName(asset.getName() == null ? asset.getId() : asset.getName());
+            }
```

Comment:
I believe the logic needs to be cleaned up a bit.

I think your lookupAsset should check if it's PROVIDING or CONSUMING, so you won't need the if here anymore